src/components/Scaler/index.js

```diff
@@ -83,53 +100,67 @@ class Scaler extends Component {
-        };
-        subDomainsByItemId[item.id] = {
-          [Axes.time]: this.props.dataContext.timeSubDomain || [
-            ...(stripPlaceholderDomain(
-              Axes.time(this.state.subDomainsByItemId[item.id])
-            ) || item.timeSubDomain),
-          ],
-          [Axes.x]: [
-            ...(stripPlaceholderDomain(
-              Axes.x(this.state.subDomainsByItemId[item.id])
-            ) || item.xSubDomain),
-          ],
-          [Axes.y]: [
-            ...(stripPlaceholderDomain(
-              Axes.y(this.state.subDomainsByItemId[item.id])
-            ) || item.ySubDomain),
-          ],
-        };
-      };
-      (this.props.dataContext.series || []).forEach(updateDomainsForItem);
-      (this.props.dataContext.collections || []).forEach(updateDomainsForItem);
+    const changedSeries = findItemsWithChangedDomains(
+      prevProps.dataContext.series,
+      this.props.dataContext.series
+    );
+    const changedCollections = findItemsWithChangedDomains(
+      prevProps.dataContext.collections,
+      this.props.dataContext.collections
+    );
+    if (changedSeries.length > 0 || changedCollections.length > 0) {
+      const domainsByItemId = { ...this.state.domainsByItemId };
+      const subDomainsByItemId = { ...this.state.subDomainsByItemId };
+
+      []
+        .concat(changedSeries)
+        .concat(changedCollections)
+        .forEach(item => {
+          domainsByItemId[item.id] = {
+            [Axes.time]: this.props.dataContext.timeDomain || [
+              ...(item.timeDomain ||
+                stripPlaceholderDomain(
+                  Axes.time(this.state.domainsByItemId[item.id])
+                )),
+            ],
+            [Axes.x]: [
+              ...(item.xDomain ||
+                stripPlaceholderDomain(
+                  Axes.x(this.state.domainsByItemId[item.id])
+                ) ||
+                PLACEHOLDER_DOMAIN),
+            ],
+            [Axes.y]: [
+              ...(item.yDomain ||
+                stripPlaceholderDomain(
+                  Axes.y(this.state.domainsByItemId[item.id])
+                ) ||
+                PLACEHOLDER_DOMAIN),
+            ],
+          };
+          subDomainsByItemId[item.id] = {
+            [Axes.time]: this.props.dataContext.timeSubDomain || [
+              ...(item.timeSubDomain ||
+                stripPlaceholderDomain(
+                  Axes.time(this.state.subDomainsByItemId[item.id])
+                )),
+            ],
+            [Axes.x]: [
+              ...(item.xSubDomain ||
+                stripPlaceholderDomain(
+                  Axes.x(this.state.subDomainsByItemId[item.id])
+                ) ||
+                PLACEHOLDER_DOMAIN),
+            ],
+            [Axes.y]: [
+              ...(item.ySubDomain ||
+                stripPlaceholderDomain(
+                  Axes.y(this.state.subDomainsByItemId[item.id])
+                ) ||
+                PLACEHOLDER_DOMAIN),
+            ],
```

Comment:
This one is huge, would you mind to rewrite it into a separate unified function?

Reply:
I worry that splitting this out will actually make it more confusing to read. When they're all side-by-side like this, the repetition is obvious, and you'll notice if something is doing something special.

Thoughts?

Reply:
I believe unification is stands for not to observe `something is doing something special`, but I don't have any strong opinion here, so it's up to you.